workspaces/adventure-pack/goodies/typescript/Iterator.from/index.ts

```diff
@@ -0,0 +1,54 @@
+import "../Iterator.prototype.toIterable";
+
+declare global {
+  interface IteratorConstructor {
+    from<T>(
+      object: Iterator<T> | Iterable<T> | { next(): IteratorResult<T> },
+    ): IterableIterator<T>;
+  }
+
+  const Iterator: IteratorConstructor;
+}
+
+(globalThis as Record<string, unknown>).Iterator ??= {};
+
+(globalThis as Record<string, unknown>).Iterator = {
+  from<T>(
+    object: Iterator<T> | Iterable<T> | { next(): IteratorResult<T> },
+  ): IterableIterator<T> {
+    if (typeof (object as Iterable<T>)[Symbol.iterator] === "function") {
+      const iterator = (object as Iterable<T>)[Symbol.iterator]();
+      if (typeof iterator === "object" && Symbol.iterator in iterator) {
+        return iterator as IterableIterator<T>;
+      } else {
+        return {
+          ...iterator,
+          [Symbol.iterator]() {
+            return this;
+          },
+        } as IterableIterator<T>;
+      }
```

Comment:
Can we use the implementation of `.toIterable` here? I'm hoping we can call it with a custom `this`, similar to what we were discussing this weekend.

```suggestion
      return iteratorPrototype.toIterable.call(iterator);
```

Reply:
I'm a bit confused did we want the custom annotation kind of like

```TypeScript
declare global {
  interface IteratorConstructor {
    from<T>(
      this: Iterator<T>, // this context
      object: Iterator<T> | Iterable<T> | { next(): IteratorResult<T> },
    ): IterableIterator<T>;
  }

  const Iterator: IteratorConstructor;
}
```    

Also while returning
`return iteratorPrototype.toIterable.call(iterator);`
it seems that I need to assert `as IterableIterator<T>` because of the type error

```
Type 'IterableIterator<unknown>' is not assignable to type 'IterableIterator<T>'.
  The types returned by 'next(...)' are incompatible between these types.
    Type 'IteratorResult<unknown, any>' is not assignable to type 'IteratorResult<T, any>'.
      Type 'IteratorYieldResult<unknown>' is not assignable to type 'IteratorResult<T, any>'.
        Type 'IteratorYieldResult<unknown>' is not assignable to type 'IteratorYieldResult<T>'.
          Type 'unknown' is not assignable to type 'T'.
            'T' could be instantiated with an arbitrary type which could be unrelated to 'unknown'.ts(2322)
```